src/dbnode/storage/index.go

```diff
@@ -524,8 +539,15 @@ func (i *nsIndex) writeBatches(
 				batch.MarkUnmarkedEntryError(m3dberrors.ErrTooPast, idx)
 				return
 			}
+
+			if i.forwardIndexDice.forwardIndexing(entry.Timestamp) {
+				forwardWriteEntry := entry
+				forwardWriteEntry.Timestamp = entry.Timestamp.Add(blockSize)
+				forwardWriteBatch.Append(forwardWriteEntry, d)
+			}
 		})
 
+	batch.AppendAll(forwardWriteBatch)
```

Comment:
Unfortunately this isn't safe as is. When trying to index a Series, we track the last successful block starts when it was indexed (in *lookup.Entry via the shard map); and further, maintain a reference count to Series entry. Both of these are updated via  the `OnIndexSeries` per element in the WriteBatchEntry.

By just appending entries the way it's being done right now, it's going to definitely mess up the reference count (because you'll dec for each copy, even though we only have a inc on the original). 

Won't necessarily mess up the last successful write times but probably need to consider some nuances there too.

Reply:
Re: reference counting - we can either inc the appropriate references when making copies (safest), or ensure we only dec the final copy amongst all the ones we make here (can't be the original, has to be last one we track because of cleanup semantics in the Shard). 

Re: tracking last successful write time update. Here's a [good explanation](https://github.com/m3db/m3/blob/master/src%2Fdbnode%2Fstorage%2Fseries%2Flookup%2Fentry.go#L145-L153). 

AFAICT, the changes you're introducing can make it so that we're writing to 4 block starts (instead of the 3 we currently assume worst case). I'm not a 100% sure that's accurate tho. Should work that out, and if so it makes sense to bump the size of the static alloc from 3 to 4.